homeassistant/components/flo/device.py

```diff
@@ -0,0 +1,156 @@
+from aioflo.errors import RequestError
+from async_timeout import timeout
+
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN as FLO_DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FloDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """Flo device object."""
+
+    def __init__(
+        self, hass: HomeAssistantType, api_client: API, location_id: str, device_id: str
+    ):
+        """Initialize the device."""
+        self.hass: HomeAssistantType = hass
+        self.api_client: API = api_client
+        self._flo_location_id: str = location_id
+        self._flo_device_id: str = device_id
+        self._manufacturer: str = "Flo by Moen"
+        self._device_information: Optional[Dict[str, Any]] = None
+        self._water_usage: Optional[Dict[str, Any]] = None
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{FLO_DOMAIN}-{device_id}",
+            update_interval=timedelta(seconds=60),
+        )
+
+    async def _async_update_data(self):
+        """Update data via library."""
+        try:
+            async with timeout(10):
+                await asyncio.gather(
+                    *[self._update_device(), self._update_consumption_data()]
+                )
+        except (RequestError) as error:
+            raise UpdateFailed(error)
+
+    @property
+    def location_id(self) -> str:
+        """Return Flo location id."""
+        return self._flo_location_id
+
+    @property
+    def id(self) -> str:
+        """Return Flo device id."""
+        return self._flo_device_id
+
+    @property
+    def device_name(self) -> str:
+        """Return device name."""
+        return f"{self.manufacturer} {self.model}"
+
+    @property
+    def manufacturer(self) -> str:
+        """Return manufacturer for device."""
+        return self._manufacturer
+
+    @property
+    def mac_address(self) -> str:
+        """Return ieee address for device."""
+        return self._device_information["macAddress"]
+
+    @property
+    def model(self) -> str:
+        """Return model for device."""
+        return self._device_information["deviceModel"]
+
+    @property
+    def rssi(self) -> float:
+        """Return rssi for device."""
+        return self._device_information["connectivity"]["rssi"]
+
+    @property
+    def last_heard_from_time(self) -> str:
+        """Return lastHeardFromTime for device."""
+        return self._device_information["lastHeardFromTime"]
+
+    @property
+    def device_type(self) -> str:
+        """Return the device type for the device."""
+        return self._device_information["deviceType"]
+
+    @property
+    def available(self) -> bool:
+        """Return True if device is available."""
+        return self._device_information["isConnected"] and self.last_update_success
```

Comment:
You should turn this around. If the first update fails, `device_information` might be `None`

Reply:
done